# Ridder Game of Dice
https://fivethirtyeight.com/features/can-you-score-some-basketball-tickets/

In [1]:
import math

from itertools import combinations, product
from pprint import pprint

## Planning

Define a role to be a tuple of length four, e.g. (6,3,5,1), equivalent to a single roll of four dice. Create a list called `rolls` of all possible roll sets. This list should have length 6**4 = 1296.

There are 6 possible sums of two dice for each roll, in this example the sums are (9,11,7,8,4,6). Create a list of these roll sum sets. This list should have length 1296.

Then create a list of all possible guess sets, where each guess set has four numbers and the numbers range from 2 to 12. There should be 11 choose 4 combinations in this set, so 330.

Then for each guess set, loop through the roll sum sets and check if any of the guesses is in each roll sum set. Calculate the percentage of roll sum sets each guess set is in. The guess set with the highest percentage is the best guess.

In [2]:
math.comb(11,4)

330

## Work

### Create a list of all possible rolls

In [3]:
rolls = list(product(range(1,7), repeat = 4))

In [4]:
rolls[0:10]

[(1, 1, 1, 1),
 (1, 1, 1, 2),
 (1, 1, 1, 3),
 (1, 1, 1, 4),
 (1, 1, 1, 5),
 (1, 1, 1, 6),
 (1, 1, 2, 1),
 (1, 1, 2, 2),
 (1, 1, 2, 3),
 (1, 1, 2, 4)]

In [5]:
len(rolls)

1296

### Create a list of sums for each roll

In [6]:
sums = [[sum(combo) for combo in combinations(roll, r = 2)] for roll in rolls]

In [7]:
sums[:10]

[[2, 2, 2, 2, 2, 2],
 [2, 2, 3, 2, 3, 3],
 [2, 2, 4, 2, 4, 4],
 [2, 2, 5, 2, 5, 5],
 [2, 2, 6, 2, 6, 6],
 [2, 2, 7, 2, 7, 7],
 [2, 3, 2, 3, 2, 3],
 [2, 3, 3, 3, 3, 4],
 [2, 3, 4, 3, 4, 5],
 [2, 3, 5, 3, 5, 6]]

In [8]:
len(sums)

1296

### Create a list of all possible guesses

In [9]:
guesses = list(combinations(range(2,13), r = 4))

In [10]:
guesses[0:10]

[(2, 3, 4, 5),
 (2, 3, 4, 6),
 (2, 3, 4, 7),
 (2, 3, 4, 8),
 (2, 3, 4, 9),
 (2, 3, 4, 10),
 (2, 3, 4, 11),
 (2, 3, 4, 12),
 (2, 3, 5, 6),
 (2, 3, 5, 7)]

In [11]:
len(guesses)

330

In [12]:
def correct_guess_percentage(guess, sums):
    """Calculates the percent of sum sets that contain one or more numbers from a guess"""
    correct_guess = tuple(any(n in roll_sum for n in guess) for roll_sum in sums)
    return sum(correct_guess) / len(sums)

In [13]:
# Create a list of tuples containing the guess and the correct guess percentage
guess_results = [(guess, correct_guess_percentage(guess, sums)) for guess in guesses]

In [14]:
# Sort the results
sorted_guess_results = sorted(guess_results, key=lambda guess: guess[1], reverse=True)

In [15]:
pprint(sorted_guess_results[0:10])

[((4, 6, 8, 10), 0.9753086419753086),
 ((2, 6, 8, 10), 0.9614197530864198),
 ((4, 6, 8, 12), 0.9614197530864198),
 ((4, 6, 7, 9), 0.9552469135802469),
 ((5, 7, 8, 10), 0.9552469135802469),
 ((4, 7, 8, 9), 0.9537037037037037),
 ((5, 6, 7, 10), 0.9537037037037037),
 ((4, 6, 7, 8), 0.9529320987654321),
 ((4, 6, 7, 10), 0.9529320987654321),
 ((4, 7, 8, 10), 0.9529320987654321)]


## Results
I get that the best guess is (4,6,8,10) and should result in a win 97.5% of the time. I'm somewhat surprised that it doesn't contain 7, which is the most frequent combination of two dice. I think this is because we are choosing four numbers and so it is beneficial to spread out your guesses.